src/h3lib/lib/algos.c

```diff
@@ -677,36 +673,207 @@ void H3_EXPORT(polyfill)(const GeoPolygon* geoPolygon, int res, H3Index* out) {
-    bboxCenter(&bboxes[0], &center);
-    H3Index centerH3 = H3_EXPORT(geoToH3)(&center, res);
-
-    // From here on it works differently, first we get all potential
-    // hexagons inserted into the available memory
-    H3_EXPORT(kRing)(centerH3, minK, out);
-
-    // Next we iterate through each hexagon, and test its center point to see if
-    // it's contained in the GeoJSON-like struct
-    for (int i = 0; i < numHexagons; i++) {
-        // Skip records that are already zeroed
-        if (out[i] == 0) {
-            continue;
+
+    // Get the estimated number of hexagons and allocate some temporary memory
+    // for the hexagons
+    int numHexagons = H3_EXPORT(maxPolyfillSize)(geoPolygon, res);
+    // This algorithm assumes that the number of vertices is usually less than
+    // the number of hexagons, but when it's wrong, this will keep it from
+    // failing
+    const Geofence geofence = geoPolygon->geofence;
+    int numVerts = geofence.numVerts;
+    for (int i = 0; i < geoPolygon->numHoles; i++) {
+        numVerts += geoPolygon->holes[i].numVerts;
+    }
+    if (numHexagons < numVerts) numHexagons = numVerts;
+    H3Index* search = calloc(numHexagons, sizeof(H3Index));
+    H3Index* found = calloc(numHexagons, sizeof(H3Index));
+
+    // Some metadata for tracking the state of the search and found memory
+    // blocks
+    int numSearchHexes = 0;
+    int numFoundHexes = 0;
+
+    // 1. Get all points of the main polygon, convert them to hexagons and add
+    // them to the search hash. The hexagon containing the geofence point may or
+    // may not be contained by the geofence (as the hexagon's center point may
+    // be outside of the boundary.)
+    for (int i = 0; i < geofence.numVerts; i++) {
+        GeoCoord origin, destination;
+        origin = geofence.verts[i];
+        destination = i == geofence.numVerts - 1 ? geofence.verts[0]
+                                                 : geofence.verts[i + 1];
+        const int numHexesEstimate =
+            lineHexEstimate(&origin, &destination, res);
+        for (int j = 0; j < numHexesEstimate; j++) {
+            GeoCoord interpolate;
+            interpolate.lat =
+                (origin.lat * (numHexesEstimate - j) / numHexesEstimate) +
+                (destination.lat * j / numHexesEstimate);
+            interpolate.lon =
+                (origin.lon * (numHexesEstimate - j) / numHexesEstimate) +
+                (destination.lon * j / numHexesEstimate);
+            H3Index pointHex = H3_EXPORT(geoToH3)(&interpolate, res);
+            // A simple hash to store the hexagon, or move to another place if
+            // needed
+            int loc = (int)(pointHex % numHexagons);
+            int loopCount = 0;
+            while (found[loc] != 0) {
+                if (loopCount > numHexagons) {
+                    free(search);
+                    free(found);
+                    free(bboxes);
+                    return -1;
+                }
```

Comment:
The only uncovered LOC are the safeguards against if the `found` array is 100% full of non-duplicated `H3Index`es. This is where the error code comes up (there are three of these throughout the function) because it's theoretically possible, but I cannot see any way to actually trigger it, since we've created a pretty gigantic memory buffer equal to the theoretical maximum number of hexagons in the bounding box of the polygon, while this is only dealing with an outline of the polygon, which should always be less-than-or-equal-to the bounding box hexagon count.

@isaacbrodsky what are you thoughts? I believe the code would function perfectly without these checks and then we could make the function signature `void` again, but I don't want to potentially cause an infinite loop and I think having any API function with a `void` signature was, in hindsight, a bad idea, since it has restricted our ability to call out errors (for instance -- what if an issue was found with the current polyfill that we couldn't reasonably address except via an error message? We'd still have to break the API in that case to add it).

Reply:
This is why I want to add a way for effectively all functions to return errors. :)

Other than crashing or changing the signature, the other options I can see involve either returning partial output, or returning no output (zero the buffer).

There are some error conditions that have been excluded from coverage, along with an argument for why they are unreachable. I think whether we should add that exclusion would depend on how convincing the argument that the code is unreachable can be made. It does seem convincing that the containing shape (bounding box) should always have an equal or larger perimeter in hexagons than the contained shape (polygon).

Reply:
Oh, it's not even bounding box perimeter versus shape perimeter, it's a bounding box *filled* with hexagons versus the shape perimeter. The reason I did that is because it's not actually reasonable to assume what you just said -- if there's a polygon that snakes back and forth across the whole bounding box, it's perimeter could be defined by all of the hexagons within the bounding box if perfectly aligned to snake around the hexagon centerpoints. That seems to be the upper limit.

The only way to reach this particular error condition should be via a bug in the code itself, which I don't want to say is impossible and why I don't want to eliminate this guard or the error code.

...but I also don't want this PR to, well, languish until we're ready to do a 4.0 release, since that's been in discussion for over a year, since it solves a real issue (thanks to @nrabinowitz for the initial prototyping) and gets a 2-3x perf improvement over the current version.

Should we take a `kRing`/`hexRing` approach to this and use this algo first and then fall back to the old one if an error occurs? It will bloat up the codebase and I almost guarantee that the fallback will never be hit, but the output memory object will remain oversized compared to this, as well.

Reply:
Would we be able to test it? In principle that sounds OK that we would use a fallback in situations this is unable to handle (until 4.x anyways)

Reply:
Since I don't think it's possible to trigger, not naturally, no, but we could have a test-only hook that causes a "failure" and confirms the fallback is called and will return the right thing.

I'm not sure which is the least bad way to do it, but currently leaning towards a macro that is true in release builds and false in debug builds.

Reply:
I'd suggest against keying on the Debug/Release build type. I would rather the library did not change behavior based on whether it's under development.

It might be better to use a kRing like approach, and cover the backup approach in its own tests, but leave these lines uncovered.

Reply:
My thoughts:

- I don't think we want this to offer a backwards-incompatible API. That would potentially break all the bindings, including those managed by outside contributors, and would make upgrading more difficult.

- I'd prefer not to keep the old approach unless we really need to - it adds code complexity and could become a maintenance burden. As @dfellis notes, we'd also be required to allocate significantly more memory than this function now requires.

- Proposal: Could we make this `polyfill` a static/internal function, and export a wrapper that zeroes out the `out` array if we hit an error code? Then the core `polyfill` is ready for v4, we have an error fallback if needed, but we have a backwards-compatible API for v3.

Reply:
I like this suggestion. @isaacbrodsky your thoughts?

There is the very-small-but-still-nonzero chance that this will cause `polyfill` to return an empty array that it wouldn't currently do, but we can always revive the old code if needed or figure out why it errored and fix it at that time.

Reply:
I think returning a zeroed array is a reasonable option. It will require binding authors to check for this and possibly throw an "empty polyfill" exception so we should note this in the changelog. There is also the possibility that a user is using polyfill on polygons that return no indexes, this case will not be differentiable until 4.x.